governance/delegation-interop.md

```diff
@@ -0,0 +1,65 @@
+
+# Proposed Solution
+
+## Rationale
+
+Due to the fact that the `GovernanceDelegation` contract now has awareness of the block numbers and chain IDs of all of the voting power checkpoints each L2 can now be responsible for
+their own checkpointing and thus make as many updates necessary without cost to relayers. From an implementation standpoint many of the changes are quite small and should minimize the
+risk of complication of an already intricate interface.
+
+The main goal is to ensure that the incentives are set up correctly so that relayers and indexers can manage and provide users the latest delegation changes when exercising their voting
+power.
+
+## Implementation
+
+Again, the implementation involves updates the Governor that will ensure that for voting power that has been migrated can be retrieved and updated based on the state on each of the L2 chains. It accomplishes this by keeping track of the token nonce and block numbers of each checkpoint in the mainnet contract. Instead of directly sending messages between the different instances, Governors on non-OP mainnet chains will now emit events regarding these blocks in the their checkpoint updates.
+
+The new storage values will be used to keep track of both the voting power that has been accumulated on the native chain as well as others thus not breaking backwards compatibility for
+non-migrated tokens. Indexers will now need to be aware of the events being emitted on both OP mainnet and the other L2's that implement the `GovToken` to show checkpoints that have yet
+to be processed on the mainnet contract. Each L2 will now have the total cumulative voting power for every partial delegation. Vote retrieval on mainnet will take in a parameter that specifies the height of the latest block it cares about of a given L2 chain.
+
+
+## Invariants
+
+The implementation should maintain the following invariants:
+
+1. **Awareness of migration status**: To ensure that existing voting power is preserved that do not have L1 block number.
+2. **Checkpoint mapping**: A new storage field that contains the block number and L2 chain IDs of the latest checkpoint.
+3. **Nonce on token transfers**: The governance token will now employ a nonce on each transfer that will be checked in the `afterTokenTransfer` hook.
+
+# Risks & Uncertainties
+
+1. The biggest risk in implementing this change is ensuring that all existing voting power is preserved such that the checkpoints that already exist on the contract is accessible.
+2. Indexers must now accept a delay in updates if wishing to give an accurate voting status for a particular delegate.
+3. Minimize the griefing vector of incurring costs onto delegators.
+4. Estimation of costs of operating relayers to ensure they are still incentivized to relay messages.
+5. Batch ordering of checkpoints across all L2 chains.
```

Comment:
Hmmm not sure what this means, can you elaborate?

Reply:
Perhaps this means to batch process the checkpoints on each L2 to reduce frequency & cost of updates (especially when multiple small transfers occur). At least I think that is what it means, and would like to see that as part of the design. :) I feel like this may be tricky though since batching could delay the update to `Governor` on mainnet maybe?

Reply:
Yes this would be exposing some batching function to process the messages in the cross message passing Inbox